REST Matomo Tracking integration for Bitstream views

vins01-4science opened the pull request:
> [!IMPORTANT]
>  _Please note that this work has been funded by **University of Maryland**_

## References
* Related to https://github.com/openaire/OpenAIRE-Piwik-DSpace/issues/3
* Related to https://github.com/DSpace/DSpace/issues/8414
* Related to DSpace/dspace-angular#4018

> [!TIP]
>  _I would suggest to do a functional review of  #10435 ( this PR ) and DSpace/dspace-angular#4018 together._

## Description
### What change is being made?
Add `Matomo` analytics integration into `DSpace`:
- event handlers
- client interfaces
- enrichers 
- configuration files

### Why are these changes being made?
This integration allows DSpace to track and record UsageEvents with Matomo Analytics, leveraging its capabilities for enhanced tracking and analysis.

There are two main parts of this development, the rest part is responsible for tracking `Bitstream` downloads, and links them with the `angular` integration that is responsible of sending the events for the site navigation, `Item` view and search.

With this implementation we will be able to track the user interactions with the site inside `Matomo` analytics.

![image](https://github.com/user-attachments/assets/f9b71926-07e6-4802-b22f-1869578b7a57)
> In **Green** general information of the tracked user
> In **Red** you can find the requests made by the angular integration ( item view and search )
> In **Blue** you can find the requests made by the rest integration ( bitstream download )

## Instructions for Reviewers
### Prerequisites
1. Have a Matomo instance - you can use a free tier online, or you could just install it using the provided `docker-compose-matomo.yml` file, like this: (**Required Docker**)
    ```
    docker compose -p d9 -f docker-compose.yml -f dspace/src/main/docker-compose/docker-compose-matomo.yml up -d
    ```
2. Once started, you should be able to visit the [Matomo Home Page](http://localhost:8081/)
3. On the Matomo configuration you should:
  - Create a proper user
  - Create a siteId  - [Matomo-Docker-SiteId](http://localhost:8081/index.php?module=SitesManager&action=index&idSite=1&period=day&date=yesterday) and save it property.
  - Create a new authToken - [Matomo-Docker-AuthToken](http://localhost:8081/index.php?module=UsersManager&action=addNewToken&idSite=1&period=day&date=yesterday) and save it properly ( be careful since this token shouldn’t set to allow only secured requests otherwise it won't work )
4. Configure the following properties ( by replacing the one with the `${}` format with real values obtained previously ) :
  ```cfg
  # Enables matomo integration mapped inside the MatomoEventListener
  matomo.enabled = true
  # Configured `siteid` inside the matomo dashboard
  matomo.request.siteid = 1
  # The tracking endpoint of matomo
  matomo.async-client.baseurl = ${matomo-url}/matomo.php
  # Optional The auth-token configured inside the matomo instance
  matomo.async-client.token = ${matomo-auth-token}
  ```
5. Startup DSpace and try to download a bitstream that has been placed inside the `ORIGINAL` bundle ( default bundle configured for `Matomo` integration )
6. You should see a mapped request inside your `Matomo` dashboard with the bitstream details, something like the screenshot in the details of this PR - [Mattomo-Docker-Dashboard](http://localhost:8081/index.php)

---

### Configurations
<details><summary><code>matomo.cfg</code></summary>

Inside this configuration file you can setup the basic information for `Matomo` integration

```cfg
# Enables matomo integration mapped inside the MatomoEventListener
matomo.enabled = false
# Configured `siteid` inside the matomo dashboard
matomo.request.siteid = 1
# Specifies bitstream's bundle that will be tracked ( default is ORIGINAL )
# Add 'none' to disable the tracking for bitstreams
# matomo.track.bundles = ORIGINAL

# Configuring the synchronous client
# The tracking endpoint of matomo
matomo.default-client.baseurl = http://localhost:8081/matomo.php
# The auth-token configured inside the matomo instance
matomo.default-client.token =

# Configuring the ascynchronous client
# The tracking endpoint of matomo
matomo.async-client.baseurl = http://localhost:8081/matomo.php
# The auth-token configured inside the matomo instance
matomo.async-client.token =
```

</details> 

<details><summary><code>matomo.xml</code></summary>

Inside this configuration file you can customize the `Matomo` integration by changing beans.

This first portion specifies how the request should be composed, depending on the `enrichers` configured, and also identifies the `siteId` to be used.

```xml
<!-- This is the Builder that can be customized with different enrichers -->
<bean id="matomoRequestDetailsBuilder" class="org.dspace.matomo.model.MatomoRequestDetailsBuilder">
    <!-- These are the enrichers configured, each of them is responsible to create a parameter for the request starting from the UsageEvent recieved -->
    <constructor-arg name="enrichers">
        <list>
            <bean id="userAgentEnricher" class="org.dspace.matomo.factory.MatomoRequestDetailsEnricherFactory" factory-method="userAgentEnricher"/>
            <bean id="actionNameEnricher" class="org.dspace.matomo.factory.MatomoRequestDetailsEnricherFactory" factory-method="actionNameEnricher"/>
            <bean id="urlEnricher" class="org.dspace.matomo.factory.MatomoRequestDetailsEnricherFactory" factory-method="urlEnricher"/>
            <bean id="downloadEnricher" class="org.dspace.matomo.factory.MatomoRequestDetailsEnricherFactory" factory-method="downloadEnricher"/>
            <bean id="ipEnricher" class="org.dspace.matomo.factory.MatomoRequestIpAddressEnricher"/>
            <bean id="requestCountryEnricher" class="org.dspace.matomo.factory.MatomoRequestCountryEnricher"/>
            <bean id="cookieIdEnricher" class="org.dspace.matomo.factory.MatomoRequestCookieIdentifierEnricher"/>
        </list>
    </constructor-arg>
    <constructor-arg name="siteId" value="${matomo.request.siteid:}"/>
</bean>
```

This second portion declares one of the `MatomoSyncEventHandler`  that will be notified by the `EventListener`.
You can customize it by changing the `matomoClient` and the `builder` of the request.

```xml
<bean id="matomoSyncEventHandler" class="org.dspace.matomo.MatomoSyncEventHandler">
    <constructor-arg name="matomoClient" ref="matomoAsyncClient"/>
    <constructor-arg name="builder" ref="matomoRequestDetailsBuilder"/>
</bean>
```
> This configurable bean allows to have multiple `matomo` instances with different parameters.
</details> 

<details><summary><code>event-service-listeners.xml</code></summary>

Here you can find the `MatomoEventListener` that receives the `UsageEvent` and dispatches it through the configured `EventHandler`s.

```xml
<!-- Matomo events recording -->
<bean class="org.dspace.matomo.MatomoEventListener">
    <property name="eventService" ref="org.dspace.services.EventService"/>
</bean>
```

</details> 


### Changes
List of changes in this PR:
* Created a proper `Handler` to track the `UsageEvent`
* Refactored `GoogleAsyncEventHandler` just to extract common methods
* Created a basic `MatomoAbstractClient` that is responsible to format the `MatomoRequestDetails` and send it using its internal `httpClient`
* Created two different `MatomoAbstractClient` that can handle request synchronously or asynchronously. The synchronous client can be configured to ship bulk requests all at once.
* Created classes to store request details, values and response
* Created different `Enricher`s that can be configured to extract values from `UsageEvent` to enrich the request that will be sent to `Matomo`
* Created `Converter`s and `Reader`s to handle JSON data.
* Created a proper module configuration file `matomo.cfg`
* Placed all spring configurations inside  `matomo.xml` file and added listener inside the `event-service-listeners.xml`
* Created proper Tets to verify the behavior of these changes

## Checklist

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [ ] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

vins01-4science committed:
[CST-18963] Adds Matomo HTTP API integration

vins01-4science committed:
[CST-18963] Fixes configuration and adds async client for matomo REST endpoint

vins01-4science committed:
[CST-18963] Minor fixes after PR review.

vins01-4science committed:
[CST-18963] Adds JavaDocs.

vins01-4science committed:
[CST-18963] Adds missing header inside MatomoSyncEventHandlerTest

vins01-4science committed:
[CST-18963] Adds matomo enricher related to the cookie that identifies a visitor

vins01-4science committed:
[CST-18963] Refactors matomo event handler to track bitstream view

vins01-4science committed:
[CST-18963] Fixes checksyle issue

dsteelma-umd commented:
I do not have much experience with Matomo, so apologies if some of this is off-base, or otherwise obvious.

I tested with a stock DSpace 9 (based on DSpace commit hash 7131b7a9e5 and dspace-angular commit hash 97312820d6), and in general the functionality appears to work as advertised.

I am running the DSpace 9 instance on a remote server (not in a local Docker setup), and communicating with the actual Matomo website, in a configuration which closely matches our existing DSpace installation.

Some things I've noticed when testing this issue:

## 1) Matomo tracking does not start immediately after accepting the cookie popup

When accessing the website in a private browser window, the cookie popup is displayed, as expected. After clicking "That's ok" to allow the cookies, no data is sent to Matomo, even if the user navigates to other pages. The page must be manually reloaded before data is sent to Matomo.

In using a regular (non-private) browser window, the same behavior occurs on the very first visit to the website (i.e., no data is sent to Matomo after accepting the cookies popup, unless the user manually reloads the page). If the user exits the web browser and comes back to the website, then Matomo data is sent normally.

This occurs in both Firefox and Chrome.

----
**Steps to Reproduce**

1) Open an private browser window and display the "Network" tab in "Developer Tools"

2) Go to the DSpace website. The cookie popup will be displayed. Left-click the "That's ok" button. Note that in the "Network" tab, no information has been sent to the Matomo.

3) Click on some items on the page. In the "Network" tab, note that no information is being sent to Matomo.

4) Manually reload the page. In the "Network" tab, note that information is now sent to Matomo.

This occurs in both Firefox and Chrome.

----

## 2) Searches are double-reported to Matomo

I’ve seen instances where a single search in the Angular front-end is double-reported to Matomo (i.e., a single search appears twice in succession in Matomo).

----
**Steps to Reproduce**

1) Go to the DSpace website. Type a search into the search textbox, such as "lorem ipsum" and left-click the "Search" button. The search results page will display (likely with no results).

2) In Matomo, there will be two "lorem ipsum" search actions (displayed with a magnifying glass icon).

This occurs in both Firefox and Chrome.

----

## 3) Bitstream downloads are reported as a different "visitor" than website actions

When navigating through the website, Matomo builds up a "visitor profile" of the user's actions, based on a visitor id. This allows all the visitor's actions to be seen as a sequence in the Matomo web interface.

When a file is downloaded, the information sent to Matomo by the DSpace back-end does not include the visitor id, and so it not associated with the other actions taken by the user. For example in the following screenshot:

![matomo_visitor_sequence](https://github.com/user-attachments/assets/7c6fbf4b-b35c-4d7c-8d93-1a480569c314)

the download (the first entry with a single action represented by the "cloud and download arrow" icon) is not associated with the second entry, which is the sequence of three actions that led to the download).

Ideally, the download would be included in the sequence of actions (i.e., that the three navigation actions and the download action would all be in the same visit, not separated into two separate visits).

## 4) In the “Instructions for Reviewers - Prerequisites - Step 3” above, marking the “matomo.async-client.token” parameter as optional seems misleading because the default configuration of the “matomoRequestDetailsBuilder” in “dspace/config/spring/rest/matomo.xml” has the “ipEnricher” and “requestCountryEnricher” enabled by default, corresponding to the Matomo Tracking API “cip” and “country” parameters, respectively.

According to https://developer.matomo.org/api-reference/tracking-api, the “cip” and “country” parameters require authentication via “token_auth”. If a “token_auth” is not provided, bitstream tracking in the DSpace backend fails with the following error in the log:

```
Unable to track requestBody: {"requests":["?idsite=<SITE_ID>&rec=1&apiv=1&download=<URL>&action_name=<PAGE_TITLE>&cip=<IP ADDRESS>&ua=<USER_AGENT>&url=<URL>]}, response was: {"status":"success","tracked":0,"invalid":1}
```

and the bitstream download is *not* recorded by Matomo.

Therefore, in the default configuration, a Matomo auth token is *required*, not optional.

The workaround, if you don’t want to use a Matomo auth token, is to comment out the following lines in the “matomoRequestDetailsBuilder” in “dspace/config/spring/rest/matomo.xml”:

```
    <bean id="matomoRequestDetailsBuilder" class="org.dspace.matomo.model.MatomoRequestDetailsBuilder">
        <constructor-arg name="enrichers">
            <list>
                ...
                <!-- <bean id="ipEnricher" class="org.dspace.matomo.factory.MatomoRequestIpAddressEnricher"/> -->
                <!-- <bean id="requestCountryEnricher" class="org.dspace.matomo.factory.MatomoRequestCountryEnricher"/> -->
                ...
            </list>
    </bean>
```

vins01-4science commented:
Hello @dsteelma-umd, thank you for this review.
 
I would like to resolve all of these issues.

For points 1) and 2) it would be better if you report them directly inside the  https://github.com/DSpace/dspace-angular/pull/4018 PR, since they seem to be related with the angular implementation.

For point 3) - Did you accept the cookie on the website?
For point 4) -  You are completely right, those two parameters are allowed only if the authentication token has been set. Since we are sending that request from the REST server, to have a proper tracking we must set that authentication token ( otherwise those information will be always the same and related directly to the REST server ). What do you think @tdonohue?

dsteelma-umd commented:
@vins01-4science - I have copied comments 1 and 2 into the https://github.com/DSpace/dspace-angular/pull/4018 PR, as requested.

For the third item, I did accept the cookie popup on the website.

In fact (this might be an additional issue) the download is recorded even if the user declines.  For example:

1) Go to the DSpace website in a private window. The cookie popup will be displayed. Left-click the "Decline" button.

2) Select an item, and download a bitstream associated with the item. In the Matomo dashboard, note that information about the download is displayed, even though the user has declined the Matomo Analytics cookie.

**Expectations:** That since the user declined in the cookie popup, that information about the download will not be sent to Matomo.

I don't know how serious and issue it is that Matomo records the download even after the user has requested that we not track. (I am not a lawyer, and I do not know what the GPDR may or may not require).

artlowel reviewed (commented):
Thanks @vins01-4science! 

I've tested this PR while reviewing the front end PR, and tracking page views works.

I wasn't able to find any download events in my matomo dashboard though. I see the pagviews on the `/bitstreams/${bitstream_id}/download` page in the UI, but not actual download events sent from the backend. 

Now that may be due to my lack of experience with Matomo UI, or a configuration error. Because if I set a breakpoint in the MatomoEventListener, I does look like events get sent from the dspace side. So if other reviewers are able to get download events to work, I'm sure it's fine

vins01-4science committed:
[CST-18963][#10435] Adds custom cookie tracking and enrichers for Matomo requests

vins01-4science commented:
Thank you both @artlowel  and @dsteelma-umd  for your reviews.

I made some changes to disable by default the two parameters that would need auth-token: `ipEnricher` and `requestCountryEnricher`.
Moreover, now the matomo tracking feature will be allowed only if the user accepts the cookie, so only we receive a proper cookie `_pk_id` from the angular side.
I placed three more enrichers that can extract the matomo cookies from the request received and map them into proper parameters / cookie that would be sent to the matomo tracking endpoint.

Let me know if you find something else.

vins01-4science committed:
[CST-18963][#10435] Fix test failures due to missing cookie on request

dsteelma-umd commented:
@vins01-4science - I have tested your latest changes, and I also am not seeing a Matomo entry for the download of the bitstream from the DSpace backend (which I was seeing previous to these changes).

Based on some logging that I inserted into the code, it looks like the "MatomoEventListener" class is not sending the event to the event handlers because the cookie are not being found.

Do changes need to be made to the Angular frontend to actually send the cookie in the request? I am currently using the changes up to Git hash 601c6cd from https://github.com/DSpace/dspace-angular/pull/4018

vins01-4science committed:
Merge remote-tracking branch 'DSpace/main' into task/main/CST-18963

dsteelma-umd commented:
@vins01-4science - I did accept the Matomo cookie, and I do see the Angular frontend activity in Matomo.

In the Developer console on the page containing the download link, I see the following cookies:

```
_pk_id.12.62a7:"19c10e4f93f7d237.1741799314."
_pk_ref.12.62a7:"%5B%22%22%2C%22%22%2C1741799314%2C%22%2Fhome%22%5D"
_pk_ses.12.62a7:"1"
CORRELATION-ID:"b09128a2-b974-46e9-8a30-5e7111727f56"
dsLanguage:"en"
orejime-anonymous:"{"authentication":true,"preferences":true,"acknowledgement":true,"matomo":true}"
XSRF-TOKEN:"44accb3d-2353-4525-9399-4e31314355fc"
```

Once redirected to the backend and the downloaded file is displayed in the browser, the only cookie shown in the Developer console is:

```
DSPACE-XSRF-COOKIE:"44accb3d-2353-4525-9399-4e31314355fc"
```

I tried doing this in both normal and private browser windows, in Firefox, and Chrome.

I am running this in a production-like environment, where the backend and frontend have different URLs (notionally, https://api.dspace.foo.bar.baz.edu/server and https://dspace.foo.bar.baz.edu/ respectively).

When I do the following steps:

1) Go to the DSpace home page, and left-click "That's ok" in the cookie notification popup

2) On the DSpace home page, left-click the topmost item in the "Recent Submissions" list. The item page is displayed.

3) On the item page, left-click the download link of the file in the "Files" section of the left sidebar. The browser redirects to the backend and displays the file (in my case, a PDF).

In Matomo, I see three events:

* The visit to the home page
* The visit to the item page
* The click on the link to the file

All these events are shown using the "folder" icon in Matomo

What I'm expecting is to see a fourth event, with a "download" icon (i.e., a "cloud" with a downward-pointing arrow) that comes from the DSpace backend.

I tested using your "4Science:task/main/CST-18963" branch merged into the DSpace "main" branch at Git commit 7ef694f03a, and for the frontend, the "4Science:task/main/CST-18964" branch merged into the "dspace-angular" branch at Git commit b9e164de78.

Hope this either elucidates the problem, or shows you what I'm doing incorrectly.

Thanks.

tdonohue reviewed (changes requested):
@vins01-4science : Thanks for this work!  Overall, this looks great!  I tested it, and the basics work well.  

**However, I do see the same that @dsteelma-umd describes above.**  Page views are all tracked perfectly (and correctly stop tracking if I change my cookie settings in Orejime).  However, I also do not see any download events (from the backend) tracked in Matomo. I only see the page views reported from the frontend.  For example, here's what the Matomo "Visit Logs" show when I visit a Collection, go to an Item page & then download a PDF.  As you can see in the screenshot, the `/download` "page view" occurs, but I don't see the download of the PDF itself in Matomo.

![matomo-logs](https://github.com/user-attachments/assets/e5d4e237-11f3-4eff-ba79-0c38850d0994)

One other suggestion I have is that I **find the Matomo `docker-compose.yml` VERY USEFUL** for testing.  Personally, I'd recommend we add it directly to a `[src]/dspace/src/main/docker-compose/docker-compose-matomo.yml` file.  That way, it's much easier to spin up a Matomo instance for testing alongside the backend by doing something like this:
```
docker compose -p d9 -f docker-compose.yml -f dspace/src/main/docker-compose/docker-compose-matomo.yml up -d
```
You could also then copy your basic Matomo testing instructions (from this GitHub PR) into the [README](https://github.com/DSpace/DSpace/blob/main/dspace/src/main/docker-compose/README.md) in that directory, similar to how we have testing instructions for [Shibboleth in that README](https://github.com/DSpace/DSpace/blob/main/dspace/src/main/docker-compose/README.md#run-dspace-8-rest-and-shibboleth-sp-in-apache-from-your-branch).  You could add a similar section for DSpace + Matomo right after those Shibboleth instructions if you wish.

I just feel that this Docker testing environment will be **very useful** in the future if we want to debug or enhance Matomo integrations in any way.

Finally, I also did a code review.  The code looks great overall.  Most of my inline comments are requests for JavaDocs or minor questions. Basically, this looks great overall.

vins01-4science commented:
@tdonohue Thank you for your review, I am going to work on it as soon as possible.

vins01-4science commented:
As requested here: https://github.com/DSpace/dspace-angular/pull/4018#pullrequestreview-2679569394 by @tdonohue , some matomo properties should be exposed and retrieved directly from the UI.

paulo-graca commented on dspace-api/src/main/java/org/dspace/matomo/MatomoAsyncDequeHandler.java:
```diff
@@ -0,0 +1,76 @@
+ *
+ * @author Vincenzo Mecca (vins01-4science - vincenzo.mecca at 4science.com)
+ **/
+public class MatomoAsyncDequeHandler extends MatomoAbstractHandler {
```
I agree... since it's described as to be used for bulk operations, i suggest to have `bulk` in the class naming.
Perhaps: `MatomoAsyncBulkRequestHandler`

paulo-graca commented on dspace/config/spring/rest/matomo.xml:
```diff
@@ -0,0 +1,53 @@
+    </bean>
+
+    <bean id="matomoSyncEventHandler" class="org.dspace.matomo.MatomoSyncEventHandler">
+        <constructor-arg name="matomoClient" ref="matomoAsyncClient"/>
```
It would help to have documentation for this file. I'm unable to understand why, for for a sync event handler we have the AsyncClient. Why can't we use the default client? Is it intend to have different clients here, could he have all options listed here (but under comments)?

vins01-4science replied:
The event handler has its own business logic, collects events and then sends them to the `matomoClient` bean configured, in order to communicate those events to our tracking service.
This client can have its own business logic, so should not be coupled with the event-handler.
With this approach you can use all the combinations you prefer, depending also on your infrastructure.

paulo-graca replied:
Thank you @vins01-4science for addressing my feedback!

paulo-graca reviewed (commented):
I think XML documentation would help who is implementing this and how to choose the intended configuration.

vins01-4science committed:
[CST-18963] Adds docker-compose-matomo details

vins01-4science committed:
[CST-18963] PR Review

vins01-4science committed:
[CST-18963] PR Review

vins01-4science committed:
[CST-18963] Fixes Test failure

vins01-4science committed:
[CST-18963] Adds Tests for MatomoRequestTrackerIdentifierParamEnricher

vins01-4science committed:
[CST-18963] Adds Tests for MatomoAbstractClient

vins01-4science committed:
[CST-18963] Fixes checkstyle issue

vins01-4science committed:
[CST-18963] Fixes checkstyle issue

vins01-4science commented:
Hello @tdonohue , all the feedbacks have been addressed.
I found out that the cookies that i was searching for, were not sent by the `matomo-client` implementation on the angular application ( those cookies are not configured to be CROSS-DOMAIN ). 
To address this problem I've introduced a new request parameter `trackerId` that will contain the identifier of the tracker that will be used to correctly track the user actions.
This parameter will be provided only if the matomo tracker is enabled on the FE side.

Moreover, I've exposed the properties you asked for in the angular issue. 
These properties are:
- `matomo.enabled`
- `matomo.request.siteid`
- `matomo.tracker.url`


Lastly, I've added the docker-compose file for the matomo dev container.

Let me know if you find something else. Thank you.

vins01-4science committed:
Merge branch 'main' into task/main/CST-18963

# Conflicts:
#	dspace/config/dspace.cfg

artlowel reviewed (changes requested):
Thanks @vins01-4science!

I can verify that new properties work and are used from the UI, and that downloads are tracked properly now

However, if I don't agree to matomo tracking in the cookie popup in the UI, my downloads are still tracked. To reproduce:
- Open the UI in an incognito window.
- Click customize on the cookie popup, uncheck matomo and click save
- Navigate to an item with a bitstream and download it

Result:
- the page views leading up to the download don't show up in matomo, the download itself does

tdonohue commented:
@vins01-4science and @artlowel : I think we are OK to track anonymously if it's possible to do so **without adding a cookie**.  In the cookie settings you are opting out of **tracking cookies**, but you are not opting out of all statistical information you may be generating by using the system.

So, as long as the current behavior does NOT add any Matomo tracking cookies, then I feel it's OK if the download is tracked anonymously.  I haven't had a chance to test it myself, but planned to do so later today

tdonohue commented on dspace/config/modules/matomo.cfg:
```diff
@@ -0,0 +1,39 @@
+# Add 'none' to disable the tracking for bitstreams
+# matomo.track.bundles = ORIGINAL
+# Specifies custom cookies that will be tracked ( default is _pk_ref,_pk_hsr,_pk_ses ) directly from the request
+matomo.request.customcookies.cookie-name = _pk_ref,_pk_hsr,_pk_ses
```
Because this configuration doesn't appear to be used in `dspace/config/spring/rest/matomo.xml` (it's commented out), I'd recommend commenting it out here as well. 

I also admit, I don't understand this setting, and what cookies these default settings are referencing.  So, it may need more documentation at some point.

vins01-4science replied:
This settings are related to the cookies that can be generated on the Angular side.
However they're not usable since the cookies are not sent in CROSS-DOMAIN environments, and are disabled by default.
I can comment them out and add more details ( like I've done inside the matomo.xml file)

tdonohue reviewed (changes requested):
@vins01-4science : I gave this another test today.  I'm still having difficulty seeing **any downloads** in Matomo.  Is there an extra configuration required to track downloads?

Here's an example of what I see in Matomo's "Visits Log" when I search for "kitten", then click on an result, and download its file.  If you notice, there's no *download* icon appearing, just folder icons:
![downloads-matomo](https://github.com/user-attachments/assets/767968be-316b-45cc-b08b-37d2fe6defbc)

Is this the expected behavior?  Or are my downloads not tracking for some reason?  I expected to see a reference to the actual bitstream download (e.g. a URL that looked like: `http://localhost:8080/server/api/core/bitstreams/[:uuid]/content`)

Beyond that oddity, everything else looks good overall.  When I turn off Matomo tracking cookies in the Cookie Settings, I don't see any further tracking. 

One other minor issue is that with the Matomo cookie settings turned *off* (i.e. no tracking allowed), I see the requests for download have an empty `trackerId` parameter like this:
```
http://localhost:8080/server/api/core/bitstreams/2fd640bc-3225-4618-9060-e34a161b4ecd/content?trackerId=
```

Instead of passing an empty value, I'd recommend **removing** the `trackerId` parameter as the empty parameter does nothing.

One other minor thing inline below.

tdonohue commented:
@vins01-4science : I wasn't able to see any of the download icons (the little cloud with the down arrow).  As I noted, the only actions I'm seeing are searches (magnifying glass icon) & views (folder icon).  I'm running everything using the Docker compose script provided in this PR:
```
docker compose -p d9 -f docker-compose.yml -f dspace/src/main/docker-compose/docker-compose-matomo.yml up -d
```

Then I setup Matomo in that Docker image, and configure the `matomo.default-client.token` properly with the token generated by Matomo.

I'm running the User Interface in production mode (`npm run build:prod; npm run serve:ssr`).

Any ideas of what I could be doing wrong or how to debug this?

vins01-4science commented:
> @vins01-4science : I wasn't able to see any of the download icons (the little cloud with the down arrow). As I noted, the only actions I'm seeing are searches (magnifying glass icon) & views (folder icon). I'm running everything using the Docker compose script provided in this PR:
> 
> ```
> docker compose -p d9 -f docker-compose.yml -f dspace/src/main/docker-compose/docker-compose-matomo.yml up -d
> ```
> 
> Then I setup Matomo in that Docker image, and configure the `matomo.default-client.token` properly with the token generated by Matomo.
> 
> I'm running the User Interface in production mode (`npm run build:prod; npm run serve:ssr`).
> 
> Any ideas of what I could be doing wrong or how to debug this?

The problem is the internal network of the container, we should access the matomo instance using its network alias ( i.e. matomo ) instead of using the default configuration ( localhost ).

The port should be the internal one ( 80 ), I think we can add it as env property like it has been done for solr and postgresql.

```
      # matomo.tracker.url: Ensure we are using the 'matomo' image for Matomo
      matomo__P__tracker__P__url: http://matomo:8081
```

vins01-4science committed:
[CST-18963] Fixes issues with matomo on docker compose

feat:
  - Adds matomo.tracker.url override inside the docker-compose.
  - Improves logging of exceptions during connection for the matomo client

tdonohue commented:
@vins01-4science : I'm trying this again today with your latest Docker changes and Matomo and am running into new issues.  The issue I'm having is that the `matomo.tracker.url` needs to be different values for client side code versus server-side code.

Here's how I normally test PRs.
1. First, I build the backend in Docker, and start it up via Docker.  So, for this PR, I've built the PR and am running `docker compose -p d9 -f docker-compose.yml -f dspace/src/main/docker-compose/docker-compose-matomo.yml up -d` to start the backend.
2. For the frontend, I usually try to test in *production mode*.  So, I do *not* use the Docker scripts, but instead build/start the frontend manually via `npm install && npm run build:prod && npm run serve:ssr`.   I setup the frontend to point at http://localhost:8080/server/, which is where the backend is running via Docker.

This all works well for most PRs.  But, for this Matamo PR, it won't work because the frontend & backend appear to need different values of `matomo.tracker.url`.  
* For the *backend* to send download data to Matomo, the `matomo.tracker.url` must be set to the value you've set (`http://matomo`), as that lets the backend in Docker communicate with Matomo (also in Docker).
* For the *frontend* though, the URL of `http://matomo` **doesn't work** because the frontend attempts to download Matomo via `http://matomo/matomo.js` which returns a 404 error.  Instead, the frontend needs to use `http://localhost:8081/matomo.js`

This issue almost seems similar to the one we've had related to communication with the **backend** via Public or Private URLs (see https://github.com/DSpace/dspace-angular/pull/3358).

I'm trying to figure out the best way to work around this in my local environment.  I'm not yet sure whether this is an issue that could show up in some Production scenarios, but we should consider whether we need the ability to (optionally) set a different `tracker.url` for the Frontend, if the backend one isn't sufficient.

tdonohue commented:
@vins01-4science : Regarding my prior comment.  One "solution" (if we feel this should be possible in Production) could be to restore the `matomo.trackerUrl` setting you originally offered for `config.*.yml` in #4018, but make it *optional*.  If it is NOT specified, then the UI should use the backend's `matomo.tracker.url`.  But, if it *is* specified, the the UI will use it's own separate `trackerUrl` to communicate with Matomo.

That seems like it'd fix the problem I'm encountering.  But, not sure if you or @artlowel feel that'd be a bad idea to allow the frontend & backend to communicate with Matomo using different URLs.

artlowel commented:
In my earlier tests I ran only matomo in docker, and the backend locally, and that worked.

@tdonohue asked me to test the command
```
docker compose -p d9 -f docker-compose.yml -f dspace/src/main/docker-compose/docker-compose-matomo.yml up -d
```
to run both the backend and matomo in docker.

And indeed, I get the same issue he does: the UI can't reach `http://matomo` 

>That seems like it'd fix the problem I'm encountering. But, not sure if you or @artlowel feel that'd be a bad idea to allow the frontend & backend to communicate with Matomo using different URLs.

That's fine by me, as long as it's not the default

vins01-4science commented:
> That seems like it'd fix the problem I'm encountering. But, not sure if you or @artlowel feel that'd be a bad idea to allow the frontend & backend to communicate with Matomo using different URLs.

For me it's okay. However I think that we should think a common standardized way to handle this kind of things.
It will be really counterintuitive, even error-prone, to set up this kind of difference for the same thing between REST and Angular.
I am going to do this quick fix and test it again.

vins01-4science committed:
Merge remote-tracking branch 'origin/main' into task/main/CST-18963

# Conflicts:
#	dspace/config/modules/rest.cfg

tdonohue reviewed (approved):
👍 Thanks @vins01-4science !  I was able to test & review this again today. All my feedback has been addressed.  Additionally, it now works perfectly in Docker with the addition of the frontend's (optional) `matomo.trackerUrl`.

1. Running `docker compose -p d9 -f docker-compose.yml -f dspace/src/main/docker-compose/docker-compose-matomo.yml up -d` to startup backend & Matomo in Docker
2. Verified if I do NOT change the `matomo.trackerUrl` on the frontend, then the frontend will use the backend's `matomo.tracker.url = http://matomo`, which results in the prior issue of failing to download http://matomo/matomo.js from the frontend.
3. But, now, I add this to the `config.*.yml` and **everything works perfectly**:
```
matomo:
  trackerUrl: 'http://localhost:8081'
```
4. The frontend will access Matomo via http://localhost:8081, so tracking of page hits & searches works correctly
5. The backend will continue to access Matomo via http://matomo, so the tracking of *downloads* now works properly as well.

Regarding @artlowel 's prior feedback about whether to track downloads anonymously even if you *reject* the Matomo Cookie settings.  I think your current approach is fine for now.  I don't see it as a privacy issue because the download data seems anonymous.

I do understand that the anonymous download info may look odd to some sites, but I feel they may rather know the *actual number of downloads* instead of excluding download data from users that reject the Matomo cookie.

I also did verify that if you reject the Matomo tracking cookie, then you are not tracked in anything else you do (page hits, searches, etc).  Only your downloads will appear, but without identifying information.

Overall, this looks good to me.  @artlowel : any final thoughts on this?  Are you OK with it as-is? (We could always fix it later if we find that users do *not* like the anonymous download behavior)

artlowel reviewed (approved):
It is ok with me

tdonohue commented:
Merging with +2 votes.  Thanks again @vins01-4science !  Please remember that we do need Matomo Documentation added to the 9.0 Docs at https://wiki.lyrasis.org/display/DSDOC9x/.  Therefore, I'm flagging this as `needs documentation` until proper documentation has been added.
